components/feature/recentlyclosed/src/main/java/mozilla/components/feature/recentlyclosed/RecentlyClosedTabsStorage.kt

```diff
@@ -0,0 +1,121 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.recentlyclosed
+
+import android.content.Context
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.browser.session.ext.writeState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.feature.recentlyclosed.adapter.RecentlyClosedTabAdapter
+import mozilla.components.feature.recentlyclosed.db.RecentlyClosedTabEntity
+import mozilla.components.feature.recentlyclosed.db.RecentlyClosedTabsDatabase
+import java.io.File
+import java.util.UUID
+
+/**
+ * A storage implementation that saves snapshots of recently closed tabs / sessions.
+ */
+@Suppress("TooManyFunctions")
+class RecentlyClosedTabsStorage(
```

Comment:
In newer code we start to move away from exposing storage classes directly to apps. Instead we now let the middleware load the state (e.g. the list of recently closed tabs) and add them to `BrowserState` by dispatching actions on the store. When dispatching actions (e.g. to remove a recently closed tab from the list) the middleware will see that and do it in the storage too. So from the app's point of view it only interacts with `BrowserStore` and doesn't need to deal with storage at all. This pattern works pretty well for small amounts of data, which I assume the list of recently closed tabs will be.

An example of this is `feature-containers`.

Is this something that would work here too? With the middleware already saving closed tabs it feels like we are already almost there. :)